src/tests/suite/daos_checksum.c

```diff
@@ -330,10 +338,57 @@ io_with_server_side_verify(void **state)
 			     &ctx.update_iod, &ctx.update_sgl, NULL);
 	assert_int_equal(rc, -DER_CSUM);
 	cleanup_cont_obj(&ctx);
+	unset_csum_fi();
+
+	/**5. Data corruption. Update should fail due CRC mismatch */
+	set_client_data_corrupt_fi();
+	setup_cont_obj(&ctx, DAOS_PROP_CO_CSUM_CRC64, false, 0, OC_SX);
+	rc = daos_obj_update(ctx.oh, DAOS_TX_NONE, 0, &ctx.dkey, 1,
+			     &ctx.update_iod, &ctx.update_sgl, NULL);
+	assert_int_equal(rc, 0);
+	cleanup_cont_obj(&ctx);
 
+	setup_cont_obj(&ctx, DAOS_PROP_CO_CSUM_CRC64, true, 0, OC_SX);
+	rc = daos_obj_update(ctx.oh, DAOS_TX_NONE, 0, &ctx.dkey, 1,
+			     &ctx.update_iod, &ctx.update_sgl, NULL);
+	assert_int_equal(rc, -DER_CSUM);
+	cleanup_cont_obj(&ctx);
 	unset_csum_fi();
+	cleanup_data(&ctx);
+}
+
+static void
+test_server_data_corruption(void **state)
+{
+	test_arg_t	*arg = *state;
+	struct csum_test_ctx	 ctx = {0};
+	int			 rc;
+
+	setup_from_test_args(&ctx, *state);
+	setup_cont_obj(&ctx, DAOS_PROP_CO_CSUM_CRC64, false, 1024*8, OC_SX);
+
+	/**1. Simple server data corruption after RDMA */
+	setup_multiple_extent_data(&ctx);
+	/** Set the Server data corruption flag */
+	rc = daos_mgmt_set_params(arg->group, -1, DMG_KEY_FAIL_LOC,
+				DAOS_CHECKSUM_SDATA_CORRUPT | DAOS_FAIL_ALWAYS,
+				0, NULL);
+	assert_int_equal(rc, 0);
+	/** Perform the update */
+	rc = daos_obj_update(ctx.oh, DAOS_TX_NONE, 0, &ctx.dkey, 1,
+			&ctx.update_iod, &ctx.update_sgl, NULL);
+	assert_int_equal(rc, 0);
+	/** Clear the fail injection flag */
+	rc = daos_mgmt_set_params(arg->group, -1, DMG_KEY_FAIL_LOC, 0, 0, NULL);
+	assert_int_equal(rc, 0);
+	/** Fetch should result in checksum failure : SSD bad data*/
+	rc = daos_obj_fetch(ctx.oh, DAOS_TX_NONE, 0, &ctx.dkey, 1,
+			    &ctx.fetch_iod, &ctx.fetch_sgl, NULL, NULL);
+	assert_int_equal(rc, -DER_CSUM);
```

Comment:
Just curious,  once we fully support checksum, fetch suppose to return 0 and correct data?

Reply:
This test case corrupts data on servers. So we should get checksum error.

There is another test case data is corrupted on one server and fetch will perform the correct data from the replicas (or fallback server).
	/* 5. Replicated object with corruption /
	if (test_runable(*state, 2)) {
		cleanup_cont_obj(&ctx);
		setup_cont_obj(&ctx, DAOS_PROP_CO_CSUM_CRC64, false, 1024*8,
			       OC_RP_2GX);
		rc = daos_obj_update(ctx.oh, DAOS_TX_NONE, 0, &ctx.dkey, 1,
				     &ctx.update_iod, &ctx.update_sgl, NULL);
		assert_int_equal(rc, 0);
		set_fetch_csum_fi();
		rc = daos_obj_fetch(ctx.oh, DAOS_TX_NONE, 0, &ctx.dkey, 1,
				    &ctx.fetch_iod, &ctx.fetch_sgl, NULL, NULL);
		assert_int_equal(rc, 0);
		/* Update/Fetch data matches /
		assert_memory_equal(ctx.update_sgl.sg_iovs->iov_buf,
				    ctx.fetch_sgl.sg_iovs->iov_buf,
				    ctx.update_sgl.sg_iovs->iov_buf_len);

Reply:
ah, ok